src/engine_filters/pauseable.rs

```diff
@@ -0,0 +1,81 @@
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::Duration;
+
+use Sample;
+use Source;
+
+/// Filter that allows another thread to pause the stream.
+#[derive(Clone, Debug)]
+pub struct Pauseable<I> where I: Source, I::Item: Sample {
+    input: I,
+
+    //Local storage of the paused value.  Allows us to only check the remote occasionally.
+    local_paused: bool,
+
+    //The paused value which may be manipulated by another thread.
+    remote_paused: Arc<AtomicBool>,
+
+    //The frequency with which local_paused should be updated by remote_paused
+    update_frequency: u32,
+
+    //How many samples remain until it is time to update local_paused with remote_paused.
+    samples_until_update: u32,
+}
+
+impl<I> Pauseable<I> where I: Source, I::Item: Sample {
+    pub fn new(source: I, remote_paused: Arc<AtomicBool>, update_ms: u32) -> Pauseable<I> {
+        let update_frequency = (update_ms * source.get_samples_rate())/1000;
```

Comment:
In theory the samples rate of the source can change while it is playing.
However this never happens in practice and making this correct needs a lot of work, so please just add a comment here saying something like `// TODO: handle the fact that the samples rate can change`.